src/Symfony/Component/HttpClient/RetryHttpClient.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\Response\AsyncContext;
+use Symfony\Component\HttpClient\Response\AsyncResponse;
+use Symfony\Component\HttpClient\Retry\RetryBackOffInterface;
+use Symfony\Component\HttpClient\Retry\RetryDeciderInterface;
+use Symfony\Contracts\HttpClient\ChunkInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Automatically retries the failing requests.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class RetryHttpClient implements HttpClientInterface
+{
+    use AsyncDecoratorTrait;
+
+    private $decider;
+    private $strategy;
+    private $maxRetries;
+    private $logger;
+
+    private $retryCount = 0;
+
+    /**
+     * @param int $maxRetries The maximum number of times to retry
+     */
+    public function __construct(HttpClientInterface $client, RetryDeciderInterface $decider, RetryBackOffInterface $strategy, int $maxRetries = 3, LoggerInterface $logger = null)
+    {
+        $this->client = $client;
+        $this->decider = $decider;
+        $this->strategy = $strategy;
+        $this->maxRetries = $maxRetries;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        return new AsyncResponse($this->client, $method, $url, $options, function (ChunkInterface $chunk, AsyncContext $context) use ($method, $url, $options) {
+            $exception = null;
+            try {
+                // clone the chunk to let the original value of `$didthrow`
```

Comment:
I don't think this is needed. Doesn't AsyncResponse already handle this?

Reply:
issue is:
- if `$chunk->isTimeout()` throw a `TransportException`
- exception is catch above
- then, if `$retryCount >= $this->maxRetries` we abort retrying and yield the `$chunk` (that already thrown an exeption)

Reading the code of `AsyncResponse`, its looks like the chunk is not cloned. 

I may be wrong..

Reply:
ok, it's handled here: https://github.com/symfony/symfony/blob/9c8cd0827f6fab9bf840b7fd2940bcbe1b16e75e/src/Symfony/Component/HttpClient/Response/AsyncResponse.php#L336

the AsyncResponse restore the Chunk state